src/main/java/com/example/userrestapi/controller/UserController.java

```diff
@@ -23,20 +27,37 @@
 
         List<User> users = new ArrayList<User>();
 
-        if (name != null) {
+        //I get it this is probably not the most optimal solution,
```

Comment:
the easiest way here would be to get all users, as you do in line 58 and then filter it out on each parameter one by one, something like:
users = repository.findAll();
if (name != null) {
    users = users.stream().filter(user -> user.name().equals(name)).collect(Collectors.toList())
}
if (surname != null) {
...
}
etc.
optionally you can also build a stream like that:
Stream<User> userStream = users.stream();
if (name != null) {
    userStream.filter(user -> user.getName().equals(name));
}
...
and at the end collect the results and return.
To make it even prettier you can do all of this in UserRepository in a method findByParams(all params here), so this one would only contain the return. In real life world with a connection to db, this would look more like that

Reply:
so if user would have 6 parameters then findByParams() would require 2^6 ifs in it? no other option to do it? cause that's a lot of conditions 😄

Reply:
btw. I've commited fixes to the lines you've commented on. I don't know if this re-request review button works like that cause there wasn't any message about notification being sent to you and it's the first time I'm using it

Reply:
that's 6 ifs :slightly_smiling_face: you get the whole collection, they you filter it out one by one param

Reply:
the "filter" method on stream returns the collection (or a stream in a second case I described before) so you don't need to do "if x and y and z then filter and return" then "if x and y then filter and return" and so on.
You simply do filter on each param one by one on whole users array, so in each "if" you will just filter out more values (if params given) and at the end you will end up with final array containing only the Users that pass on criteria that you've given.
users = repository.findAll();
if (name != null) {
users = users.stream().filter(user -> user.name().equals(name)).collect(Collectors.toList())
}
if (surname != null) {
users = users.stream().filter(user -> user.surname().equals(surname)).collect(Collectors.toList())
}
and after those two ifs, your "users" collection will only contain users with name and surname given in param. I hope this time it's clearer, but if you have more question, ask me :slightly_smiling_face:

Reply:
alright, I get it now. thank you 🙂

Reply:
sorry to bother you, but I really can't wait to hear about the decision whether it's good or bad (hoping it's the former tho :smile:). is there a possibility someone could give me some information today?

Reply:
sorry, I didn't think about it. It's good now, I approved the pull request. Well done!